### PR TITLE
[FLINK-19221][filesystems] Introduce the LocatedFileStatus to save block location RPC requests.

### PR DESCRIPTION
**Pull request by StephanEwen**
## What is the purpose of the change

This PR helps to drastically reduce the number of RPC requests made to the file system master (for example the HDFS Name Node) when enumerating the splits in file sources. Previously, we enumerated first all files, and then made a separate RPC call to the name node to get the block locations for each file.

Hadoop has introduced an optimization here that this PR exploits. When the HDFS Client returns a `FileStatus` (description of a file) it frequently returns a `LocatedFileStatus` which already contains all the `BlockLocation` information.

In this PR, we mirror that mechanism in the Flink FileSystem APIs: 
When the `FileStatus` obtained from Hadoop file systems during listing the directory (or getting details for a file) already has the block locations, we use that directly and expose in a Flink `LocatedFileStatus`. That saves the extra RPC call per file to obtain that block location information.

## Brief change log

 - 04e493f0c0d3ba80a2507c4550633a62af45a2e3 : [FLINK-19218] Remove inconsistent/misleading locality information from Local File Splits.
    This is a general cleanup/improvement and makes the implementation of the main change easier.

  - cdc711101c86a58196f0978272a315d03b87e303 : [FLINK-19221] Main change as described above.

## Verifying this change

This change makes a "best effort" attempt and, on purpose, does not add hard tests to enforce this behavior. That is because certain file system implementations or setups may not eagerly expose the location information and we don't want to break compatibility working with them.

The best way to verify this is running a file-heavy batch job against HDFS with this change and observing the reduced number or Name Node RPC requests.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): **no**
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: **yes** *(a new class, no breaking changes)*
  - The serializers: **no**
  - The runtime per-record code paths (performance sensitive): **no**
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: **no**
  - The S3 file system connector: **no**

## Documentation

  - Does this pull request introduce a new feature? **no**
  - If yes, how is the feature documented? **not applicable**

### CODE REVIEW

File: flink-filesystems/flink-hadoop-fs/src/main/java/org/apache/flink/runtime/fs/hdfs/LocatedHadoopFileStatus.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+package org.apache.flink.runtime.fs.hdfs;
+
+import org.apache.flink.core.fs.BlockLocation;
+import org.apache.flink.core.fs.FileStatus;
+import org.apache.flink.core.fs.LocatedFileStatus;
+
+/**
+ * Concrete implementation of the {@link FileStatus} interface for the
+ * Hadoop Distribution File System.
```
Comment (aljoscha) [inline flink-filesystems/flink-hadoop-fs/src/main/java/org/apache/flink/runtime/fs/hdfs/LocatedHadoopFileStatus.java:27]:
```suggestion
 * Concrete implementation of the {@link LocatedFileStatus} interface for the
 * Hadoop Distributed File System.
```

The `Distribution` -> `Distributed` bit is a copy-and-paste error, could also fix in `HadoopFileStatus`.

File: flink-filesystems/flink-hadoop-fs/src/main/java/org/apache/flink/runtime/fs/hdfs/HadoopFileStatus.java [hunk 2]
```diff
@@ -77,4 +77,17 @@ public boolean isDir() {
...
 	}
+
+	// ------------------------------------------------------------------------
+
+	/**
+	 * Creates a new {@code HadoopFileStatus} from Hadoop's {@link org.apache.hadoop.fs.FileStatus}.
+	 * If Hadoop's file status is <i>located</i>, i.e., it contains block information, then this method
+	 * returns an implementation of Flink's {@link org.apache.flink.core.fs.LocatedFileStatus}.
+	 */
+	public static HadoopFileStatus fromHadoopStatus(final org.apache.hadoop.fs.FileStatus fileStatus) {
```
Comment (aljoscha) [inline flink-filesystems/flink-hadoop-fs/src/main/java/org/apache/flink/runtime/fs/hdfs/HadoopFileStatus.java:88]: It feels a bit off that `HadoopFileStatus` knows about its subclasses but I think here it's ok.

**Review by aljoscha — APPROVED** [review]
The changes look good to merge once CI is green! I didn't manually test if this actually reduces RPC load on the NameNode but from looking at the code it should. And the changes should be covered by existing tests, at least the behaviour from outside.

I had one typo nitpick in the code.

And one tiny nit about the commit message:
```
We her expose this on the Flink side, to save RPC calls to the name node. For example file sources often
request block locations for all files (to facilitate locality aware assignments), currently resulting in
one RPC call to the Name Node for each file.
```
should be `We here`, or sth like it. 😅

**Comment by StephanEwen** [comment]
Thanks for the review. Will fix the comments and typos.

File: flink-filesystems/flink-hadoop-fs/src/main/java/org/apache/flink/runtime/fs/hdfs/HadoopFileStatus.java (hunk 2 above)
Comment (StephanEwen) [reply flink-filesystems/flink-hadoop-fs/src/main/java/org/apache/flink/runtime/fs/hdfs/HadoopFileStatus.java:88]: The idea here was that the class itself needs not know its subclasses, but we need a factory that is aware of that.
I put the factory in that class, we could also put it into `HadoopFileSystem`. But I thought having it more visible would actually be a good idea here.

**Pull request closed without merging**